facade/service.go

```diff
@@ -66,89 +84,295 @@ func (f *Facade) AddService(ctx datastore.Context, svc service.Service) (err err
 }
 
 func (f *Facade) addService(ctx datastore.Context, svc service.Service, setLockOnCreate bool) error {
-	glog.V(2).Infof("Facade.AddService: %+v", svc)
 	store := f.serviceStore
-
-	_, err := store.Get(ctx, svc.ID)
-	if err != nil && !datastore.IsErrNoSuchEntity(err) {
+	// service add validation
+	if err := f.validateServiceAdd(ctx, &svc); err != nil {
+		glog.Errorf("Could not validate service %s (%s) for adding: %s", svc.Name, svc.ID, err)
 		return err
-	} else if err == nil {
-		return fmt.Errorf("error adding service; %v already exists", svc.ID)
 	}
-
-	// verify the service with parent ID does not exist with the given name
-	if s, err := store.FindChildService(ctx, svc.DeploymentID, svc.ParentServiceID, svc.Name); err != nil {
-		glog.Errorf("Could not verify service path for %s: %s", svc.Name, err)
+	var configFiles []servicedefinition.ConfigFile
+	if svc.ConfigFiles == nil || len(svc.ConfigFiles) == 0 {
+		for _, configFile := range svc.OriginalConfigs {
+			configFiles = append(configFiles, configFile)
+		}
+	} else {
+		for _, configFile := range svc.ConfigFiles {
+			configFiles = append(configFiles, configFile)
+		}
+	}
+	svc.ConfigFiles = nil
+	// write the service into the database
+	svc.UpdatedAt = time.Now()
+	svc.CreatedAt = svc.UpdatedAt
+	if err := store.Put(ctx, &svc); err != nil {
+		glog.Errorf("Could not create service %s (%s): %s", svc.Name, svc.ID, err)
 		return err
-	} else if s != nil {
-		err := fmt.Errorf("service %s found at %s", svc.Name, svc.ParentServiceID)
-		glog.Errorf("Cannot create service %s: %s", svc.Name, err)
+	}
+	glog.Infof("Created service %s (%s)", svc.Name, svc.ID)
+	// add the service configurations to the database
+	if err := f.updateServiceConfigs(ctx, svc.ID, configFiles, true); err != nil {
+		glog.Warningf("Could not set configurations to service %s (%s): %s", svc.Name, svc.ID, err)
+	}
+	glog.Infof("Set configuration information for service %s (%s)", svc.Name, svc.ID)
+	// sync the service with the coordinator
+	if err := f.syncService(ctx, svc.ID, setLockOnCreate, setLockOnCreate); err != nil {
+		glog.Errorf("Could not sync service %s (%s): %s", svc.Name, svc.ID, err)
 		return err
 	}
+	glog.Infof("Synced service %s (%s) to the coordinator", svc.Name, svc.ID)
+	return nil
+}
 
-	// Always add services in a stopped states so in case auto ip assignment fails,
-	// the scheduler won't spam the log complaining about missing address assignments.
-	svc.DesiredState = int(service.SVCStop)
-
-	// Strip the database version; we already know this is a create
-	svc.DatabaseVersion = 0
+func (f *Facade) validateServiceAdd(ctx datastore.Context, svc *service.Service) error {
+	store := f.serviceStore
+	// verify that the service does not already exist
+	if _, err := store.Get(ctx, svc.ID); !datastore.IsErrNoSuchEntity(err) {
+		if err != nil {
+			glog.Errorf("Could not check the existance of service %s (%s): %s", svc.Name, svc.ID, err)
+			return err
+		} else {
```

Comment:
Would we hit this on a nil error from the above `store.Get()` call?  Is that desired?